[ZEPPELIN-5559] Add note cache

Reamer opened the pull request:
### What is this PR for?
The following changes are made with this PR:
 - a note cache that holds only a certain number of notes in memory
 - a lock for each `Note`
   - No unloading of notes when in use
   - removes methods that return a Note object so that the lock is not bypassed

### What type of PR is it?
- Bug Fix/Feature

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-5559

### How should this be tested?
* CI

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes, a new configuration option has been added

Reamer commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NoteManager.java:
```diff
@@ -651,5 +667,63 @@ public void updateNotePath() {
+  /**
+   * Unloads notes when reasonable amount of notes are in loaded state.
+   * Leverage a simple LRU cache for determing evictable notes. Ensure to
+   * not evict notes during save operation (dirty).
```
In principle you are right. However, `GetNote()` always places the note at the beginning of the LRUCache. `GetNote()` is always called before information is read from a note.
As long as the concurrent executions are below the lru cache threshold, there should be no problems with concurrency.

zjffdu replied:
Sorry for late response, @Reamer I will take a look at it again tomorrow

zjffdu replied:
@Reamer I still think there are still potential issue here. Let's say the number of notes in LRUCache is equal to 50, and now there's one call of handleEviction which will evict one note (Note#unload is called). At the same time, there're 50 threads (each thread read one note) which all call NotebookServer#broadcastNote to deserialize note to frontend. So at least one of them will be affected (one note will be unload). I have to admin it is very unlikely unless very high concurrency. But in theory, this kind of case exists.

Reamer replied:
I have changed the implementation to a read/write lock. Notes that have this read or write lock are not evicted.

Reamer commented:
@zjffdu 
I have had very good experience with this feature in my environment and would like to include it in the master so that further development can be based on it.
Since I have been using this feature, Zeppelin does not run in any OOM when I look at all the notes via script.

Reamer commented:
@zjffdu 
I have changed the implementation to Read/Write-Lock and the CI is also successful. A new review would be nice.
Sorry for this big change, but protecting the Note object with a read/write lock requires big changes in several classes.

Reamer commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java:
```diff
@@ -391,13 +392,76 @@ public Note getNote(String noteId, boolean reload) throws IOException {
+      NoteBeforeSaveCallback<T> beforeSave, NoteAfterSaveCallback<T> afterSave) throws IOException {
+    final Note note = getNote(noteId, false);
+
+    if (note == null) { // callback contract: run beforeSave callback also when no note exists for given noteId
```
`NoteBeforeSaveCallback` and `NoteAfterSaveCallback` has been removed

Reamer commented:
I have found bugs in this implementation, let me correct these bugs before reviewing.

Reamer commented:
> I have found bugs in this implementation, let me correct these bugs before reviewing.

Bugs fixed.

Reamer commented:
@zjffdu 
I hope you find time between Christmas and New Year to look at the changes. I would like to add this PR to the master branch at the beginning of January. The PR drastically reduces Zeppelin's memory consumption.

zjffdu commented:
Thanks @Reamer I will try to find time to review this PR

zjffdu commented:
Sorry, @Reamer I will find time to review it by the end of this week.

zjffdu commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java:
```diff
@@ -352,43 +355,64 @@ public void removeCorruptedNote(String noteId, AuthenticationInfo subject) throw
+  public <T> T processNote(String noteId, boolean reload, NoteProcessor<T> noteProcessor) throws IOException {
+    return noteManager.processNote(noteId, reload, note -> {
+      if (note == null) {
+        return noteProcessor.process(note);
```
Is it necessary to process a note when it is null?

Reamer replied:
Yes, for several reasons.

To return a correct response as in this [case](https://github.com/apache/zeppelin/pull/4252/files#diff-017bf51cce0cfa2d227c2a93230a2164bbad6b98de581a54dc4a778b49704497R128).

```java
if (note == null) {
  return new JsonResponse<>(Response.Status.NOT_FOUND, "Note " + noteId + " not found").build();
}
```
Or to work on a callback as in this [case](https://github.com/apache/zeppelin/pull/4252/files#diff-94da03110f043520a11e2697bd4b6749134c14214ec240b65a61f6b0ed1f0005R66-R77)

```java
    return notebook.processNote(noteId,
      jobNote -> {
        List<NoteJobInfo> notesJobInfo = new ArrayList<>();
        if (jobNote == null) {
          callback.onFailure(new IOException("Note " + noteId + " not found"), context);
        } else {
          notesJobInfo.add(new NoteJobInfo(jobNote));
          callback.onSuccess(notesJobInfo, context);
        }
        return notesJobInfo;
      });
```

for logging as in this [case](https://github.com/apache/zeppelin/pull/4252/files#diff-c8d59c9816d0b26045cd0cd44343eb10dbebd61730fadae84a9b147e3b6226f8R1829-R1831)

```java
      getNotebook().processNote(noteId,
        note -> {
          if (note == null) {
            // It is possible the note is removed, but the job is still running
            LOG.warn("Note {} doesn't existed, it maybe deleted.", noteId);
          } else {
            note.clearParagraphOutput(paragraphId);
            Paragraph paragraph = note.getParagraph(paragraphId);
            broadcastParagraph(note, paragraph, MSG_ID_NOT_DEFINED);
          }
          return null;
        });
```

zjffdu commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NoteManager.java:
```diff
@@ -561,69 +569,79 @@ public String toString() {
-          note.setPath(parent.toString() + "/" + note.getName());
+    public <T> T loadAndProcessNote(boolean reload, NoteProcessor<T> noteProcessor)
+        throws IOException {
+        // load note
```
minor indention issue

Reamer replied:
adjusted

zjffdu commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NoteManager.java:
```diff
@@ -561,69 +569,79 @@ public String toString() {
 
     private Folder parent;
-    private Note note;
+    private NoteInfo noteinfo;
```
noteinfo -> noteInfo ?

Reamer replied:
adjusted

zjffdu commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NoteManager.java:
```diff
@@ -561,69 +569,79 @@ public String toString() {
         }
-        note.setCronSupported(ZeppelinConfiguration.create());
-        note.setLoaded(true);
+        note.getLock().readLock().lock();
```
Move it to the try block below seems more readable.

Reamer replied:
adjusted

zjffdu commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java:
```diff
@@ -748,4 +766,20 @@ public Boolean isRevisionSupported() {
+   * Functional Interface for note processing.
+   */
+  @FunctionalInterface
+  public static interface NoteProcessor<T> {
```
static is redundant

Reamer replied:
adjusted

zjffdu commented:
@Reamer Overall this approach LGTM, just some minor comments

Reamer commented:
I have adjusted the code based on your comments.

> @Reamer Overall this approach LGTM, just some minor comments

I have adjusted the code based on your comments. Thanks for your review.
I will merge this PR with the master branch on Friday, as long as no further comments are received.
